File: scala/ch02/ex2.5.scala
def even (n: Int) = n % 2 == 0
def square (n: Int) = n * n

def fastExpt (base: Int, n: Int): Int = n match {
    case 0 => 1
    case _ if even(n) => square(fastExpt(base, n/2))
    case _ => base * fastExpt(base, n - 1)
  }

type expPair = Int

// To find a, keep dividing z by 2, counting the number of times.
def findA (z: expPair) = {
  def f (a: Int, leftOver: expPair): Int = 
    if (leftOver % 2 == 0) f(a + 1, leftOver / 2) else a
  f(0, z)
}

// To find b, keep dividing z by 3, counting the number of times.
def findB (z: expPair) = {
  def f (b: Int, leftOver: expPair): Int = 
    if (leftOver % 3 == 0) f(b + 1, leftOver / 3) else b
  f(0, z)
}

def cons (a: Int, b: Int) = fastExpt(2, a) * fastExpt(3, b)

def car (z: expPair) = findA(z)
def cdr (z: expPair) = findB(z)

import org.scalatest._
import org.scalatest.matchers._

object expPairSpec extends Spec with ShouldMatchers {
  val zeroZero = cons(0, 0)
  val oneTwo   = cons(1, 2)
  val twoOne   = cons(2, 1)
  val fourFive = cons(4, 5)

  describe ("car on a pair, implemented as (2**a)*(3**b)") {
    it ("should return a") {
      car(zeroZero) should equal (0)
      car(oneTwo)   should equal (1)
      car(twoOne)   should equal (2)
      car(fourFive) should equal (4)
    }
  }

  describe ("cdr on a pair, implemented as (2**a)*(3**b)") {
    it ("should return b") {
      cdr(zeroZero) should equal (0)
      cdr(oneTwo)   should equal (2)
      cdr(twoOne)   should equal (1)
      cdr(fourFive) should equal (5)
    }
  }
}
expPairSpec execute

File: clojure/ch02/ex2.5.clj
(ns sicp.ch02 (:use clojure.test))
(use '[clojure.contrib.except :only (throw-if)])

;(defn even? [n] (= 0 (rem n 2)))
(defn square [x] (* x x))
(defn fast-expt [b n]
  (cond (= n 0) 1
        (even? n) (square (fast-expt b (/ n 2)))
        :else (* b (fast-expt b (- n 1)))))

; To find a, keep dividing z = (cons a b) by 2, counting the number of times.
(defn find-a [z]
  (defn f [a left-over]
    (cond (= (rem left-over 2) 0) (f (+ a 1) (/ left-over 2))
          :else a))
  (f 0 z))

; To find b, keep dividing z = (cons a b) by 3, counting the number of times.
(defn find-b [z]
  (defn f [b left-over]
    (cond (= (rem left-over 3) 0) (f (+ b 1) (/ left-over 3))
          :else b))
  (f 0 z))

(defn cons2 [a b]
  (* (fast-expt 2 a) (fast-expt 3 b)))

(defn car [z] (find-a z))
(defn cdr [z] (find-b z))

(deftest test-exp-pair
  (def zero-zero (cons2 0 0))
  (def one-two   (cons2 1 2))
  (def two-one   (cons2 2 1))
  (def four-five (cons2 4 5))
  (is (= (car zero-zero) 0))
  (is (= (cdr zero-zero) 0))
  (is (= (car one-two)   1))
  (is (= (cdr one-two)   2))
  (is (= (car two-one)   2))
  (is (= (cdr two-one)   1))
  (is (= (car four-five) 4))
  (is (= (cdr four-five) 5)))

(run-tests)
